### ISSUE TITLE
HSQLDB arrays can't be projected from nested rows

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13965]
#13872 addressed an issue where `ARRAY` types were nested in `MULTISET` in PostgreSQL. The added integration test fails in HSQLDB when trying to fetch an `ARRAY` type from a nested `ROW` projection.

It produces this query here:

```sql
select null as "nested.null",
array[] as "nested.[Ljava.lang.String;@71945bc0"
from (values(1)) as dual(dual)
```

Which looks wrong, but the problem isn't the alias. The following exception is thrown:

```
       
org.jooq.exception.DataAccessException: SQL [select null as "nested.null", array[] as "nested.[Ljava.lang.String;@71945bc0" from (values(1)) as dual(dual) -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT]; Error while reading field: row (null, array[]), at JDBC index: 1
	at org.jooq_3.18.0-SNAPSHOT.HSQLDB.debug(Unknown Source) ~[?:?]
	at org.jooq.impl.Tools.translate(Tools.java:3306) ~[classes/:?]
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:686) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1465) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1426) ~[classes/:?]
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:239) ~[classes/:?]
	at org.jooq.impl.Tools.fetchOne(Tools.java:2598) ~[classes/:?]
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:509) ~[classes/:?]
	at org.jooq.impl.Tools.attach(Tools.java:1524) ~[classes/:?]
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4662) ~[classes/:?]
	at org.jooq.impl.DefaultDSLContext.lambda$55(DefaultDSLContext.java:4682) ~[classes/:?]
	at org.jooq.impl.Tools.attach(Tools.java:1524) ~[classes/:?]
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4682) ~[classes/:?]
	at org.jooq.test.all.testcases.MultisetTests.lambda$214(MultisetTests.java:1497) ~[test-classes/:?]
	at org.jooq.test.jOOQAbstractTest.withDSLContext(jOOQAbstractTest.java:2116) ~[test-classes/:?]
	at org.jooq.test.jOOQAbstractTest.bindAndInline(jOOQAbstractTest.java:2104) ~[test-classes/:?]
	at org.jooq.test.BaseTest.bindAndInline(BaseTest.java:1760) ~[test-classes/:?]
	at org.jooq.test.all.testcases.MultisetTests.assertMultisetDataType(MultisetTests.java:1490) ~[test-classes/:?]
	at org.jooq.test.all.testcases.MultisetTests.assertMultisetDataType(MultisetTests.java:1482) ~[test-classes/:?]
	at org.jooq.test.all.testcases.MultisetTests.assertMultisetDataType(MultisetTests.java:1478) ~[test-classes/:?]
	at org.jooq.test.all.testcases.MultisetTests.testMultisetProjectionDataTypeArray(MultisetTests.java:1738) ~[test-classes/:?]
	at org.jooq.test.jOOQAbstractTest.testMultisetProjectionDataTypeArray(jOOQAbstractTest.java:5407) ~[test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413) ~[junit-4.13.2.jar:4.13.2]
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210) ~[.cp/:?]
Caused by: java.sql.SQLException: Error while reading field: row (null, array[]), at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1591) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1450) ~[classes/:?]
	... 50 more
Caused by: java.sql.SQLException: Error while reading field: "nested.null", at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1591) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1570) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1450) ~[classes/:?]
	... 50 more
Caused by: java.sql.SQLSyntaxErrorException: incompatible data type in conversion
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.hsqldb.jdbc.JDBCResultSet.getArray(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorResultSet.getArray(CursorImpl.java:466) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1320) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$InternalBinding.get(DefaultBinding.java:1030) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1581) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1570) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1450) ~[classes/:?]
	... 50 more
Caused by: org.hsqldb.HsqlException: incompatible data type in conversion
	at org.hsqldb.error.Error.error(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.hsqldb.error.Error.error(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.hsqldb.jdbc.JDBCResultSet.getArray(Unknown Source) ~[hsqldb-2.6.1.jar:2.6.1]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:698) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorResultSet.getArray(CursorImpl.java:466) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1320) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$InternalBinding.get(DefaultBinding.java:1030) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1581) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1570) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1526) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144) ~[classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1450) ~[classes/:?]
```

**Comment by lukaseder** [comment]
Seems to be a bug in HSQLDB 2.6.1. Let's see.

**Comment by lukaseder** [comment]
I've reported https://sourceforge.net/p/hsqldb/bugs/1662. We can work around this problem easily.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.5 (https://github.com/jOOQ/jOOQ/issues/13966), and 3.16.11 (https://github.com/jOOQ/jOOQ/issues/13967)
